shell/platform/linux/fl_renderer.cc

```diff
@@ -93,74 +70,85 @@ gboolean fl_renderer_setup(FlRenderer* self, GError** error) {
   }
 
   if (!eglBindAPI(EGL_OPENGL_ES_API)) {
+    EGLint egl_error = eglGetError();  // must be before egl_config_to_string()
     g_autofree gchar* config_string =
         egl_config_to_string(priv->egl_display, priv->egl_config);
     g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
                 "Failed to bind EGL OpenGL ES API using configuration (%s): %s",
-                config_string, egl_error_to_string(eglGetError()));
+                config_string, egl_error_to_string(egl_error));
     return FALSE;
   }
 
   return TRUE;
 }
 
-GdkVisual* fl_renderer_get_visual(FlRenderer* self,
-                                  GdkScreen* screen,
-                                  GError** error) {
+GdkVisual* fl_renderer_get_visual(FlRenderer* self, GdkScreen* screen) {
   FlRendererPrivate* priv =
       static_cast<FlRendererPrivate*>(fl_renderer_get_instance_private(self));
 
-  EGLint visual_id;
-  if (!eglGetConfigAttrib(priv->egl_display, priv->egl_config,
-                          EGL_NATIVE_VISUAL_ID, &visual_id)) {
-    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
-                "Failed to determine EGL configuration visual");
+  if (FL_RENDERER_GET_CLASS(self)->get_visual) {
+    return FL_RENDERER_GET_CLASS(self)->get_visual(
+        self, screen, priv->egl_display, priv->egl_config);
+  } else {
     return nullptr;
   }
+}
 
-  GdkVisual* visual =
-      FL_RENDERER_GET_CLASS(self)->get_visual(self, screen, visual_id);
-  if (visual == nullptr) {
-    g_set_error(error, fl_renderer_error_quark(), FL_RENDERER_ERROR_FAILED,
-                "Failed to find visual 0x%x", visual_id);
-    return nullptr;
+void fl_renderer_set_window(FlRenderer* self, GdkWindow* window) {
+  if (FL_RENDERER_GET_CLASS(self)->set_window) {
+    FL_RENDERER_GET_CLASS(self)->set_window(self, window);
   }
-
-  return visual;
 }
 
 gboolean fl_renderer_start(FlRenderer* self, GError** error) {
   FlRendererPrivate* priv =
       static_cast<FlRendererPrivate*>(fl_renderer_get_instance_private(self));
 
-  priv->egl_surface = FL_RENDERER_GET_CLASS(self)->create_surface(
+  EGLSurfacePair surfaces = FL_RENDERER_GET_CLASS(self)->create_surface(
```

Comment:
This function should now be called `create_surfaces`. I think it would fit better with the existing style to have two out parameters and return FALSE on error and set a GError.